File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.ArrayDeque;
import java.util.Queue;

public class Solution {
    public int solution(String begin,String target, String[] words) {
        Queue<WordState> queue = new ArrayDeque<>();
        boolean[] visited = new boolean[words.length];

        queue.offer(new WordState(begin,0));
        
        
        while (!queue.isEmpty()) {
            // 방문
            WordState curr = queue.poll();
            // 방문한 곳이 내가 찾던 target이라면 return cnt
            if(curr.word.equals(target)) {
                return curr.cnt;
            }
            
            // nextvertex 예약
            for(int i=0; i<words.length; i++) {
                if(getDiffCount(curr.word,words[i])==1) {
                    if(!visited[i]) {
                        queue.offer(new WordState(words[i],curr.cnt+1));
                        visited[i]=true;
                    }
                }
            }
            
        }
        int answer = 0;
        return answer;
    }
    int getDiffCount(String word, String target) {
        int diffCount = 0;
        for(int i=0; i<word.length(); i++) {
            if(word.charAt(i)!=target.charAt(i)) diffCount++;
        }
        return diffCount;
    }
    class WordState {
        String word;
        int cnt;

        public WordState(String word, int cnt) {
            this.word = word;
            this.cnt = cnt;
        }
    }

}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    
    visited = [False] * len(words)
    queue = deque()
    queue.append((begin,0))
    
    while queue :
        current,count = queue.popleft()
        
        if current == target :
            return count
        for i in range(len(words)) :
            if visited[i] :
                continue
                
            diff = 0
            
            for a,b in zip(current,words[i]) :
                if a!=b :
                    diff += 1
                    
            if diff == 1 :
                visited[i] = True
                queue.append((words[i],count+1))
                
    return 0
    